src/main/java/teammates/storage/sqlapi/UserDb.java

```diff
@@ -0,0 +1,223 @@
+
+    /**
+     * Gets the number of instructors created within a specified time range.
+     */
+    public long getNumInstructorsByTimeRange(Instant startTime, Instant endTime) {
+        Session session = HibernateUtil.getSessionFactory().getCurrentSession();
+        CriteriaBuilder cb = session.getCriteriaBuilder();
+        CriteriaQuery<Long> cr = cb.createQuery(Long.class);
+        Root<Instructor> root = cr.from(Instructor.class);
+
+        cr.select(cb.count(root.get("id"))).where(cb.and(
+                cb.greaterThanOrEqualTo(root.get("createdAt"), startTime),
+                cb.lessThan(root.get("createdAt"), endTime)));
+
+        return session.createQuery(cr).getSingleResult();
+    }
+
+    /**
+     * Gets the number of students created within a specified time range.
+     */
+    public long getNumStudentsByTimeRange(Instant startTime, Instant endTime) {
+        Session session = HibernateUtil.getSessionFactory().getCurrentSession();
+        CriteriaBuilder cb = session.getCriteriaBuilder();
+        CriteriaQuery<Long> cr = cb.createQuery(Long.class);
+        Root<Student> root = cr.from(Student.class);
+
+        cr.select(cb.count(root.get("id"))).where(cb.and(
+                cb.greaterThanOrEqualTo(root.get("createdAt"), startTime),
+                cb.lessThan(root.get("createdAt"), endTime)));
+
+        return session.createQuery(cr).getSingleResult();
+    }
+
+    /**
+     * Checks if an instructor exists by its {@code courseId} and {@code email}.
+     */
+    private <T extends User> boolean hasExistingInstructor(String courseId, String email) {
+        Session session = HibernateUtil.getSessionFactory().getCurrentSession();
+        CriteriaBuilder cb = session.getCriteriaBuilder();
+        CriteriaQuery<Long> cr = cb.createQuery(Long.class);
+        Root<Instructor> instructorRoot = cr.from(Instructor.class);
+
+        cr.select(cb.count(instructorRoot.get("id")))
+                .where(cb.and(
+                    cb.equal(instructorRoot.get("courseId"), courseId),
+                    cb.equal(instructorRoot.get("email"), email)));
+
+        long instructorCount = session.createQuery(cr).getSingleResult();
+
+        return instructorCount > 0;
+    }
+
+    /**
+     * Checks if a student exists by its {@code courseId} and {@code email}.
+     */
+    private <T extends User> boolean hasExistingStudent(String courseId, String email) {
+        Session session = HibernateUtil.getSessionFactory().getCurrentSession();
+        CriteriaBuilder cb = session.getCriteriaBuilder();
+        CriteriaQuery<Long> cr = cb.createQuery(Long.class);
+        Root<Student> studentRoot = cr.from(Student.class);
+
+        cr.select(cb.count(studentRoot.get("id")))
+                .where(cb.and(
+                        cb.equal(studentRoot.get("courseId"), courseId),
+                        cb.equal(studentRoot.get("email"), email)));
+
+        long studentCount = session.createQuery(cr).getSingleResult();
```

Comment:
getSingleResult throws an exception if not found. You probably want getSingleResultOrNull here.